### PR TITLE
Add linear holonomic constraint task

### PR DESCRIPTION
**Pull request by shbang91**
- added a linear holonomic constraint task class
- added an example using Draco3 robot, which has a rolling contact joint on its knee (can be modelled as a holonomic constraint) -> The robot mesh takes too much space, so just wonder if we can include this model in the robot_description repo.

### CODE REVIEW

**Comment by stephane-caron** [comment]
Great! I'll take a look at the code and give you proper feedback today.

> The robot mesh takes too much space, so just wonder if we can include this model in the robot_description repo.

Absolutely, I was going to suggest the same. The best would be to make a standalone repo for Draco3's description, then link it from ``robot_descriptions`` following [these steps](https://github.com/robot-descriptions/robot_descriptions.py/blob/master/CONTRIBUTING.md#adding-a-new-robot-description).

**Comment by stephane-caron** [comment]
One complaint from the linter: a few Black checks don't pass, you can either [check the logs](https://github.com/tasts-robots/pink/actions/runs/4227259530/jobs/7349745350) or run black locally on `linear_holonomic_task.py`.

**Comment by shbang91** [comment]
Thank you for the feedback. I've cleaned up the code based on your suggestion (I used the black). Once the pull request in the robot description repository gets accepted, I'll modify the draco3 example here and then do another pull request, if that sounds okay. 

Thank you.

**Comment by stephane-caron** [comment]
> I'll modify the draco3 example here and then do another pull request, if that sounds okay.

Making another PR without the robot description in the commit history sounds good.

Alternatively, we could continue this PR and do a single squash commit in the end. The only thing we don't want to do is merge this PR with the robot description in its commit history (otherwise the robot description will be downloaded by all future pulls of the repo).

File: pink/tasks/linear_holonomic_task.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+import numpy as np
+import pinocchio as pin
+
+from ..configuration import Configuration
+from ..utils import get_joint_idx
+from .exceptions import TargetNotSet
+from .task import Task
+
+
+class LinearHolonomicTask(Task):
```
Comment (stephane-caron) [inline pink/tasks/linear_holonomic_task.py:31]: A general linear holonomic task would be:

$$
e(q) = A q - b
$$

$$
\dot{e}(q) = A \dot{q} \Rightarrow J(q) = A
$$

In this PR what you have implemented is a particular case (where $A$ couples two joints), but we could let the user supply any matrix $A$ to implement the general linear-holonomic task.

File: pink/tasks/linear_holonomic_task.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        """
+        jacobian = self.compute_jacobian(configuration)
+        weighted_jacobian = self.cost * jacobian  # [cost]
+        H = weighted_jacobian.T @ weighted_jacobian
+        c = np.zeros(H.shape[0])
+        return (H, c)
+
+    def __repr__(self):
+        """Human-readable representation of the task."""
+        return f"HolonomicTask(cost={self.cost}, gain={self.gain})"
```
Comment (stephane-caron) [inline pink/tasks/linear_holonomic_task.py:172]:
```suggestion
        return f"LinearHolonomicTask(cost={self.cost}, gain={self.gain})"
```

File: pink/utils.py [hunk 3]
```diff
@@ -59,6 +59,23 @@ def get_root_joint_dim(model: pin.Model) -> Tuple[int, int]:
...
+
+    Returns:
+        nq: Number of configuration dimensions.
+        nv: Number of tangent dimensions.
+    """
+    if model.existJointName(joint_name):
+        joint_id = model.getJointId(joint_name)
+        joint = model.joints[joint_id]
+        return joint.idx_q, joint.idx_v
+    raise "cannot find the joint index(q, v) corresponding joint index"
```
Comment (stephane-caron) [inline pink/utils.py:76]:
```suggestion
    raise PinkError(f"cannot find the joint index corresponding to joint {joint_name}")
```
``PinkError`` is defined in ``pink.exceptions``.

File: pink/utils.py [hunk 4]
```diff
@@ -59,6 +59,23 @@ def get_root_joint_dim(model: pin.Model) -> Tuple[int, int]:
     return 0, 0
 
 
+def get_joint_idx(model: pin.Model, joint_name: str) -> Tuple[int, int]:
+    """get joint index in tangent space
```
Comment (stephane-caron) [inline pink/utils.py:63]:
```suggestion
    """Get joint index in tangent space.
```

**Review by stephane-caron — COMMENTED** [review]
Looking good. My main comment is about implementing the most general linear-holonomic task we can so that it applies to more use cases.

**Comment by shbang91** [comment]
> Alternatively, we could continue this PR and do a single squash commit in the end. The only thing we don't want to do is merge this PR with the robot description in its commit history (otherwise the robot description will be downloaded by all future pulls of the repo).
I like this idea. Thank you for bringing this up.

File: pink/tasks/linear_holonomic_task.py (hunk 1 above)
Comment (stephane-caron) [reply pink/tasks/linear_holonomic_task.py:31]: I was thinking of ``__init__(self, A, cost)``, but indeed that delegates the computation of $A$ to the user and we lose some useful code from your use case.

How about making a general `LinearHolonomicTask`, and a child

```python
class JointCouplingTask(LinearHolonomicTask):
```

That builds `self.A` internally and gives users a nicer interface?

Currently the coupling is between two joints but we could imagine a function like ``task.add_coupling(["shoulder_pitch", "elbow", "wrist"], [1.0, -0.5, 1.0])`` that the task would internally convert into adding the proper line to ``self.A``.

Comment (shbang91) [reply pink/tasks/linear_holonomic_task.py:31]: Awesome. That would work better. I'll work on that based on your suggestion.

**Pull request closed without merging**